File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/domain/bingo/BingoCard.java
package systems.zlink.samples.bingo.server.play.domain.bingo;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

public final class BingoCard {
    public static final int Size = 3;
    public static final int CellCount = Size * Size;
    public static final int FreeCellIndex = 4;

    private final List<Integer> numbers;
    private final List<Boolean> marks;

    private BingoCard(List<Integer> numbers, List<Boolean> marks) {
        this.numbers = numbers;
        this.marks = marks;
    }

    public static BingoCard from(List<Integer> submitted) {
        if (submitted == null || submitted.size() != CellCount) {
            throw new IllegalArgumentException("bingo card must contain 9 cells");
        }
        HashSet<Integer> seen = new HashSet<>();
        ArrayList<Integer> numbers = new ArrayList<>(submitted);
        for (int i = 0; i < numbers.size(); i++) {
            int number = numbers.get(i);
            if (i == FreeCellIndex) {
                if (number != 0) {
                    throw new IllegalArgumentException("bingo card center cell must be 0");
                }
                continue;
            }
            if (number < 1 || number > 15 || !seen.add(number)) {
                throw new IllegalArgumentException(
                        "bingo card numbers must be unique values from 1 to 15");
            }
        }
        ArrayList<Boolean> marks = new ArrayList<>();
        for (int i = 0; i < CellCount; i++) {
            marks.add(i == FreeCellIndex);
        }
        return new BingoCard(numbers, marks);
    }

    static BingoCard restore(List<Integer> numbers, List<Boolean> marks) {
        BingoCard card = from(numbers);
        if (marks == null || marks.size() != CellCount) {
            throw new IllegalArgumentException("bingo card marks must contain 9 cells");
        }
        return new BingoCard(new ArrayList<>(numbers), new ArrayList<>(marks));
    }

    public void markDrawnNumber(int number) {
        for (int i = 0; i < numbers.size(); i++) {
            if (numbers.get(i) == number) {
                marks.set(i, true);
            }
        }
    }

    public int completedLines() {
        int completed = 0;
        for (int row = 0; row < Size; row++) {
            completed += isRowComplete(row) ? 1 : 0;
        }
        for (int column = 0; column < Size; column++) {
            completed += isColumnComplete(column) ? 1 : 0;
        }
        completed += isDiagonalComplete(0, Size + 1) ? 1 : 0;
        completed += isDiagonalComplete(Size - 1, Size - 1) ? 1 : 0;
        return completed;
    }

    public List<Integer> numbersSnapshot() {
        return List.copyOf(numbers);
    }

    public List<Boolean> marksSnapshot() {
        return List.copyOf(marks);
    }

    private boolean isRowComplete(int row) {
        int start = row * Size;
        for (int i = 0; i < Size; i++) {
            if (!marks.get(start + i)) {
                return false;
            }
        }
        return true;
    }

    private boolean isColumnComplete(int column) {
        for (int row = 0; row < Size; row++) {
            if (!marks.get((row * Size) + column)) {
                return false;
            }
        }
        return true;
    }

    private boolean isDiagonalComplete(int start, int step) {
        for (int i = 0; i < Size; i++) {
            if (!marks.get(start + (i * step))) {
                return false;
            }
        }
        return true;
    }
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/domain/bingo/BingoCard.kt
package systems.zlink.samples.kotlin.bingo.server.play.domain.bingo

class BingoCard
private constructor(
    private val numbers: MutableList<Int>,
    private val marks: MutableList<Boolean>,
) {
    fun markDrawnNumber(number: Int) {
        numbers.forEachIndexed { index, value ->
            if (value == number) {
                marks[index] = true
            }
        }
    }

    fun completedLines(): Int {
        var completed = 0
        for (row in 0 until Size) {
            if (isRowComplete(row)) {
                completed++
            }
        }
        for (column in 0 until Size) {
            if (isColumnComplete(column)) {
                completed++
            }
        }
        if (isDiagonalComplete(0, Size + 1)) {
            completed++
        }
        if (isDiagonalComplete(Size - 1, Size - 1)) {
            completed++
        }
        return completed
    }

    fun numbersSnapshot(): List<Int> = numbers.toList()

    fun marksSnapshot(): List<Boolean> = marks.toList()

    private fun isRowComplete(row: Int): Boolean {
        val start = row * Size
        return (0 until Size).all { marks[start + it] }
    }

    private fun isColumnComplete(column: Int): Boolean =
        (0 until Size).all { row -> marks[(row * Size) + column] }

    private fun isDiagonalComplete(start: Int, step: Int): Boolean =
        (0 until Size).all { marks[start + (it * step)] }

    companion object {
        private const val Size = 3
        private const val CellCount = Size * Size
        private const val FreeCellIndex = 4

        fun from(values: List<Int>): BingoCard {
            require(values.size == CellCount) { "Bingo card must contain $CellCount cells." }
            require(values[FreeCellIndex] == 0) { "Bingo card center cell must be the free cell." }
            val seen = mutableSetOf<Int>()
            values.forEachIndexed { index, value ->
                if (index == FreeCellIndex) {
                    return@forEachIndexed
                }
                require(value in 1..15) {
                    "Bingo card value is outside the sample draw range: $value"
                }
                require(seen.add(value)) { "Bingo card contains duplicate value: $value" }
            }
            val numbers = values.toMutableList()
            val marks = MutableList(CellCount) { it == FreeCellIndex }
            return BingoCard(numbers, marks)
        }

        fun restore(values: List<Int>, marks: List<Boolean>): BingoCard {
            from(values)
            require(marks.size == CellCount) { "Bingo card marks must contain $CellCount cells." }
            return BingoCard(values.toMutableList(), marks.toMutableList())
        }
    }
}
